rcode-reader/src/main/java/eu/rssw/pct/PLReader.java

```diff
@@ -0,0 +1,198 @@
+            FileEntry fe = null;
+            while ((fe = readEntry(fc, offset, charset, version)) != null) {
+                if (fe.isValid())
+                    files.add(fe);
+                offset += fe.getTocSize();
+            }
+        } catch (IOException caught) {
+            throw new InvalidLibraryException(caught);
+        }
+    }
+
+    public boolean isMemoryMapped() {
+        return mm;
+    }
+
+    public InputStream getInputStream(FileEntry fe) throws IOException {
+        ByteBuffer bb = null;
+        try (RandomAccessFile raf = new RandomAccessFile(pl, "r")) {
+            FileChannel fc = raf.getChannel();
+            bb = ByteBuffer.allocate(fe.getSize());
+            fc.read(bb, fe.getOffset());
+        }
+
+        return new ByteArrayInputStream(bb.array());
+    }
+
+    private Charset getCharset(FileChannel fc) throws IOException {
+        ByteBuffer bEncoding = ByteBuffer.allocate(ENCODING_SIZE);
+        if (fc.read(bEncoding, ENCODING_OFFSET) != ENCODING_SIZE)
+            throw new InvalidLibraryException("Invalid charset");
+        bEncoding.position(0);
+        StringBuilder sbEncoding = new StringBuilder();
+        int zz = 0;
+        while ((zz < 20) && (bEncoding.get(zz) != 0)) {
+            sbEncoding.append((char) bEncoding.get(zz++));
+        }
+        try {
+            return Charset.forName(sbEncoding.toString());
+        } catch (IllegalArgumentException iae) {
+            return Charset.forName("US-ASCII");
+        }
+    }
+
+    private int getTOCOffset(FileChannel fc, int version) throws IOException {
+        ByteBuffer bTOC = ByteBuffer.allocate(4);
+        if (fc.read(bTOC, (version == 1 ? FILE_LIST_OFFSET : FILE_LIST_OFFSET_V11)) != 4)
+            throw new InvalidLibraryException("Unable to read table of contents");
+        return bTOC.getInt(0);
+    }
+
+    private FileEntry readEntry(FileChannel fc, int offset, Charset charset, int version)
+            throws IOException {
+        ByteBuffer b1 = ByteBuffer.allocate(1);
+        fc.read(b1, offset);
+
+        if (b1.get(0) == (byte) 0xFE) {
+            boolean stop = false;
+            int zz = 0;
+            while (!stop) {
+                b1.position(0);
+                int kk = fc.read(b1, offset + ++zz);
+                stop = (kk == -1) || (b1.get(0) == (byte) 0xFF);
+            }
+
+            return new FileEntry(zz);
+        } else if (b1.get(0) == (byte) 0xFF) {
+            b1.position(0);
+            fc.read(b1, offset + 1);
+            int fNameSize = (int) b1.get(0) & 0xFF;
+            if (fNameSize == 0)
+                return new FileEntry(29);
+            ByteBuffer b2 = ByteBuffer.allocate(fNameSize);
+            fc.read(b2, offset + 2);
+            b2.position(0);
+            String fName = charset.decode(b2).toString();
+            ByteBuffer b3 = ByteBuffer.allocate((version == 1 ? 28 : 48)); // Ou 47
+            fc.read(b3, offset + 2 + fNameSize);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Cast one of the operands of this addition operation to a "long". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS2184)